{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Featurization\n",
    "\n",
    "This notebook is intended to featurize the data associated with CRISPRi activity scores to make it easier to train a linear regression model to.\n",
    "\n",
    "The data used is that from the following paper:\n",
    "\n",
    "[Compact and highly active next-generation libraries for CRISPR-mediated gene repression and activation](https://elifesciences.org/articles/19760)\n",
    "\n",
    "__Authors__: Anthony Hein, Briana Macedo, Manya Kapoor\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  _TODO: Create similar contribution graph_\n",
    "### _TODO:  Z-standardize all numerical data__\n",
    "\n",
    "\n",
    "__Bold == Featurized!__\n",
    "\n",
    "Features available from data:\n",
    "- Distance to TSS (primary and secondary, downstream and upstream)\n",
    "- __sgRNA length__\n",
    "- __(+) or (-) stranded sgRNA__\n",
    "- __Longest consecutive run of each base__\n",
    "- __Overall nucleotide composition (% of A, C, T, G, GC, CA, AG)__\n",
    "- __Nucleotides at each position__\n",
    "- __Nucleotides dimers at each position__\n",
    "\n",
    "Features unavailable (stretch goal):\n",
    "- DNAse-seq signal\n",
    "- MNase-seq signal\n",
    "- FAIRE-seq signal\n",
    "- sgRNA overall folding free energy\n",
    "- RNA pairing at each position\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "### IMPORTS\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "### READ DFs\n",
    "activity_score_df = pd.read_csv('CRISPRi_Activity_Score_Data.csv')\n",
    "tss_annotation_df = pd.read_csv('TSS_Annotations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>PAM genomic coordinate [hg19]</th>\n",
       "      <th>strand targeted</th>\n",
       "      <th>sgRNA length (including PAM)</th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>CRISPRi activity score [Horlbeck et al., eLife 2016]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323441</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GCGCTCTGATTGGACGGAGCG</td>\n",
       "      <td>0.019320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323216</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GCCCCAGGATCAGGCCCCGCG</td>\n",
       "      <td>0.348892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323296</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GGCCGCCCTCGGAGAGCTCTG</td>\n",
       "      <td>0.912409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323318</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GACGGCGACCCTAGGAGAGGT</td>\n",
       "      <td>0.997242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323362</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GGTGCAGCGGGCCCTTGGCGG</td>\n",
       "      <td>0.962154</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene symbol chromosome  PAM genomic coordinate [hg19] strand targeted  \\\n",
       "0        AARS      chr16                       70323441               +   \n",
       "1        AARS      chr16                       70323216               +   \n",
       "2        AARS      chr16                       70323296               +   \n",
       "3        AARS      chr16                       70323318               +   \n",
       "4        AARS      chr16                       70323362               +   \n",
       "\n",
       "   sgRNA length (including PAM)         sgRNA sequence  \\\n",
       "0                            24  GCGCTCTGATTGGACGGAGCG   \n",
       "1                            24  GCCCCAGGATCAGGCCCCGCG   \n",
       "2                            24  GGCCGCCCTCGGAGAGCTCTG   \n",
       "3                            24  GACGGCGACCCTAGGAGAGGT   \n",
       "4                            24  GGTGCAGCGGGCCCTTGGCGG   \n",
       "\n",
       "   CRISPRi activity score [Horlbeck et al., eLife 2016]  \n",
       "0                                           0.019320     \n",
       "1                                           0.348892     \n",
       "2                                           0.912409     \n",
       "3                                           0.997242     \n",
       "4                                           0.962154     "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### IMPORT CHECKs\n",
    "activity_score_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS source</th>\n",
       "      <th>Primary TSS, 5'</th>\n",
       "      <th>Primary TSS, 3'</th>\n",
       "      <th>Secondary TSS, 5'</th>\n",
       "      <th>Secondary TSS, 3'</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>P1</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>58858938</td>\n",
       "      <td>58859039</td>\n",
       "      <td>58858938</td>\n",
       "      <td>58859039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>P2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>58864822</td>\n",
       "      <td>58864847</td>\n",
       "      <td>58864822</td>\n",
       "      <td>58864847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>52645379</td>\n",
       "      <td>52645393</td>\n",
       "      <td>52645379</td>\n",
       "      <td>52645393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>-</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>9268507</td>\n",
       "      <td>9268523</td>\n",
       "      <td>9268528</td>\n",
       "      <td>9268542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>chr12</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGE, matched peaks</td>\n",
       "      <td>8975206</td>\n",
       "      <td>8975223</td>\n",
       "      <td>8975144</td>\n",
       "      <td>8975169</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene transcript chromosome strand           TSS source  Primary TSS, 5'  \\\n",
       "0   A1BG         P1      chr19      -  CAGE, matched peaks         58858938   \n",
       "1   A1BG         P2      chr19      -  CAGE, matched peaks         58864822   \n",
       "2   A1CF       P1P2      chr10      -  CAGE, matched peaks         52645379   \n",
       "3    A2M       P1P2      chr12      -  CAGE, matched peaks          9268507   \n",
       "4  A2ML1       P1P2      chr12      +  CAGE, matched peaks          8975206   \n",
       "\n",
       "   Primary TSS, 3'  Secondary TSS, 5'  Secondary TSS, 3'  \n",
       "0         58859039           58858938           58859039  \n",
       "1         58864847           58864822           58864847  \n",
       "2         52645393           52645379           52645393  \n",
       "3          9268523            9268528            9268542  \n",
       "4          8975223            8975144            8975169  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_annotation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "activity_score_df['sgRNA sequence'] = activity_score_df['sgRNA sequence'].str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>PAM genomic coordinate [hg19]</th>\n",
       "      <th>strand targeted</th>\n",
       "      <th>sgRNA length (including PAM)</th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>CRISPRi activity score [Horlbeck et al., eLife 2016]</th>\n",
       "      <th>+ strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323441</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GCGCTCTGATTGGACGGAGCG</td>\n",
       "      <td>0.019320</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323216</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GCCCCAGGATCAGGCCCCGCG</td>\n",
       "      <td>0.348892</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323296</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GGCCGCCCTCGGAGAGCTCTG</td>\n",
       "      <td>0.912409</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323318</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GACGGCGACCCTAGGAGAGGT</td>\n",
       "      <td>0.997242</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323362</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GGTGCAGCGGGCCCTTGGCGG</td>\n",
       "      <td>0.962154</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene symbol chromosome  PAM genomic coordinate [hg19] strand targeted  \\\n",
       "0        AARS      chr16                       70323441               +   \n",
       "1        AARS      chr16                       70323216               +   \n",
       "2        AARS      chr16                       70323296               +   \n",
       "3        AARS      chr16                       70323318               +   \n",
       "4        AARS      chr16                       70323362               +   \n",
       "\n",
       "   sgRNA length (including PAM)         sgRNA sequence  \\\n",
       "0                            24  GCGCTCTGATTGGACGGAGCG   \n",
       "1                            24  GCCCCAGGATCAGGCCCCGCG   \n",
       "2                            24  GGCCGCCCTCGGAGAGCTCTG   \n",
       "3                            24  GACGGCGACCCTAGGAGAGGT   \n",
       "4                            24  GGTGCAGCGGGCCCTTGGCGG   \n",
       "\n",
       "   CRISPRi activity score [Horlbeck et al., eLife 2016]  + strand  \n",
       "0                                           0.019320            1  \n",
       "1                                           0.348892            1  \n",
       "2                                           0.912409            1  \n",
       "3                                           0.997242            1  \n",
       "4                                           0.962154            1  "
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Treat as categorical feature, there is a function for this ... \n",
    "activity_score_df['+ strand'] = 1*(activity_score_df['strand targeted'] == '+')\n",
    "activity_score_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Longest consecutive run of each base "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "# `base` is the nucleotide: 'A', 'C', 'T', 'G'\n",
    "# `seq` is the sequence string\n",
    "def longest_consec(seq):\n",
    "    max_count = {'A': 0, 'C': 0, 'G': 0, 'T': 0}\n",
    "    count = 0\n",
    "    current_base = None\n",
    "    for char in seq:\n",
    "        if char == current_base:\n",
    "            count += 1\n",
    "        else:\n",
    "            if current_base != None and count > max_count[current_base]:\n",
    "                max_count[current_base] = count\n",
    "            current_base = char\n",
    "            count = 1\n",
    "    return [max_count['A'], max_count['C'], max_count['G'], max_count['T']]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>PAM genomic coordinate [hg19]</th>\n",
       "      <th>strand targeted</th>\n",
       "      <th>sgRNA length (including PAM)</th>\n",
       "      <th>sgRNA sequence</th>\n",
       "      <th>CRISPRi activity score [Horlbeck et al., eLife 2016]</th>\n",
       "      <th>+ strand</th>\n",
       "      <th>Longest A Run</th>\n",
       "      <th>Longest C Run</th>\n",
       "      <th>Longest G Run</th>\n",
       "      <th>Longest T Run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323441</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GCGCTCTGATTGGACGGAGCG</td>\n",
       "      <td>0.019320</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323216</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GCCCCAGGATCAGGCCCCGCG</td>\n",
       "      <td>0.348892</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323296</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GGCCGCCCTCGGAGAGCTCTG</td>\n",
       "      <td>0.912409</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323318</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GACGGCGACCCTAGGAGAGGT</td>\n",
       "      <td>0.997242</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AARS</td>\n",
       "      <td>chr16</td>\n",
       "      <td>70323362</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GGTGCAGCGGGCCCTTGGCGG</td>\n",
       "      <td>0.962154</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18375</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>chr10</td>\n",
       "      <td>58121021</td>\n",
       "      <td>+</td>\n",
       "      <td>24</td>\n",
       "      <td>GTAATCTTAGGGATTGTGGGA</td>\n",
       "      <td>1.147518</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18376</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>chr10</td>\n",
       "      <td>58121033</td>\n",
       "      <td>+</td>\n",
       "      <td>23</td>\n",
       "      <td>GTGACTTCAAAGTAATCTTA</td>\n",
       "      <td>0.618442</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18377</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>chr10</td>\n",
       "      <td>58120760</td>\n",
       "      <td>-</td>\n",
       "      <td>24</td>\n",
       "      <td>GTGGATGTGGGGAGCGGCGAA</td>\n",
       "      <td>0.215028</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18378</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>chr10</td>\n",
       "      <td>58120818</td>\n",
       "      <td>-</td>\n",
       "      <td>24</td>\n",
       "      <td>GACACCTCAGGAACCTAAGAC</td>\n",
       "      <td>1.170726</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18379</th>\n",
       "      <td>ZWINT</td>\n",
       "      <td>chr10</td>\n",
       "      <td>58120898</td>\n",
       "      <td>-</td>\n",
       "      <td>23</td>\n",
       "      <td>GGGCACCAATCATTACACAT</td>\n",
       "      <td>0.577706</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18380 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene symbol chromosome  PAM genomic coordinate [hg19] strand targeted  \\\n",
       "0            AARS      chr16                       70323441               +   \n",
       "1            AARS      chr16                       70323216               +   \n",
       "2            AARS      chr16                       70323296               +   \n",
       "3            AARS      chr16                       70323318               +   \n",
       "4            AARS      chr16                       70323362               +   \n",
       "...           ...        ...                            ...             ...   \n",
       "18375       ZWINT      chr10                       58121021               +   \n",
       "18376       ZWINT      chr10                       58121033               +   \n",
       "18377       ZWINT      chr10                       58120760               -   \n",
       "18378       ZWINT      chr10                       58120818               -   \n",
       "18379       ZWINT      chr10                       58120898               -   \n",
       "\n",
       "       sgRNA length (including PAM)         sgRNA sequence  \\\n",
       "0                                24  GCGCTCTGATTGGACGGAGCG   \n",
       "1                                24  GCCCCAGGATCAGGCCCCGCG   \n",
       "2                                24  GGCCGCCCTCGGAGAGCTCTG   \n",
       "3                                24  GACGGCGACCCTAGGAGAGGT   \n",
       "4                                24  GGTGCAGCGGGCCCTTGGCGG   \n",
       "...                             ...                    ...   \n",
       "18375                            24  GTAATCTTAGGGATTGTGGGA   \n",
       "18376                            23   GTGACTTCAAAGTAATCTTA   \n",
       "18377                            24  GTGGATGTGGGGAGCGGCGAA   \n",
       "18378                            24  GACACCTCAGGAACCTAAGAC   \n",
       "18379                            23   GGGCACCAATCATTACACAT   \n",
       "\n",
       "       CRISPRi activity score [Horlbeck et al., eLife 2016]  + strand  \\\n",
       "0                                               0.019320            1   \n",
       "1                                               0.348892            1   \n",
       "2                                               0.912409            1   \n",
       "3                                               0.997242            1   \n",
       "4                                               0.962154            1   \n",
       "...                                                  ...          ...   \n",
       "18375                                           1.147518            1   \n",
       "18376                                           0.618442            1   \n",
       "18377                                           0.215028            0   \n",
       "18378                                           1.170726            0   \n",
       "18379                                           0.577706            0   \n",
       "\n",
       "       Longest A Run  Longest C Run  Longest G Run  Longest T Run  \n",
       "0                  1              1              2              2  \n",
       "1                  1              4              2              1  \n",
       "2                  1              3              2              1  \n",
       "3                  1              3              2              1  \n",
       "4                  1              3              3              2  \n",
       "...              ...            ...            ...            ...  \n",
       "18375              2              1              3              2  \n",
       "18376              3              1              1              2  \n",
       "18377              1              1              4              1  \n",
       "18378              2              2              2              1  \n",
       "18379              2              2              3              2  \n",
       "\n",
       "[18380 rows x 12 columns]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Longest consecutive run of each base\n",
    "longest_cons = pd.DataFrame(activity_score_df['sgRNA sequence'].apply(longest_consec).tolist(),\n",
    "                           columns = ['Longest A Run', 'Longest C Run', \n",
    "                                        'Longest G Run', 'Longest T Run'])\n",
    "pd.concat([activity_score_df, longest_cons], axis=1, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([ 314., 1687., 5072.,    0., 9882.,  884.,    0.,  307.,  147.,\n",
       "          87.]),\n",
       " array([21. , 21.7, 22.4, 23.1, 23.8, 24.5, 25.2, 25.9, 26.6, 27.3, 28. ]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAD4CAYAAAAO9oqkAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAQz0lEQVR4nO3de5CddX3H8fenRPBWJMhiMcl0aY1WtFOlKdBaWwuWa4fQGengODVjM5NqwVtrFXTGOCKdUK1UnIqTManBcUQGbckUFCOXXmYksgjlFmhSRLIQYW242DJqI9/+cX6ph+XsSXZP2LMb3q+ZnXOe3/N7zvnkmex+9nnOc86mqpAkPbv93LADSJKGzzKQJFkGkiTLQJKEZSBJAhYMO8BMHXbYYTU6OjrsGJI0b9x8880/qKqRXuvmbRmMjo4yNjY27BiSNG8k+d5U6zxNJEmyDCRJloEkib0ogyTrkzyc5I6usUOTbEqytd0ubONJcnGSbUluS3J01zYr2vytSVZ0jf96ktvbNhcnyb7+R0qS+tubI4PPAydPGjsXuLaqlgLXtmWAU4Cl7WsVcAl0ygNYDRwLHAOs3l0gbc6qru0mP5ck6Rm2xzKoqn8Bdk4aXg5saPc3AGd0jV9aHTcChyQ5AjgJ2FRVO6vqEWATcHJbd3BVfas6n5h3addjSZJmyUxfM3hJVe0AaLeHt/FFwPaueeNtrN/4eI/xnpKsSjKWZGxiYmKG0SVJk+3rF5B7ne+vGYz3VFVrq2pZVS0bGen5vglJ0gzMtAweaqd4aLcPt/FxYEnXvMXAg3sYX9xjXJI0i2b6DuSNwApgTbu9smv8nCSX0Xmx+LGq2pHkGuCvul40PhE4r6p2JvlhkuOAzcBbgU/PMJP0FKPnXjWU571vzWlDeV5pEHssgyRfAt4AHJZknM5VQWuAy5OsBO4HzmzTrwZOBbYBTwBvA2g/9M8HbmrzPlpVu1+UfgedK5aeB3ytfUmSZtEey6Cq3jzFqhN6zC3g7CkeZz2wvsf4GPDqPeWQJD1zfAeyJMkykCRZBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkCctAkoRlIEnCMpAkYRlIkrAMJElYBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJIYsAySvDfJnUnuSPKlJM9NcmSSzUm2JvlykgPb3IPa8ra2frTrcc5r4/ckOWmwf5IkabpmXAZJFgHvApZV1auBA4CzgAuBi6pqKfAIsLJtshJ4pKpeBlzU5pHkqLbdq4CTgc8kOWCmuSRJ0zfoaaIFwPOSLACeD+wAjgeuaOs3AGe0+8vbMm39CUnSxi+rqh9X1XeBbcAxA+aSJE3DjMugqh4APgHcT6cEHgNuBh6tql1t2jiwqN1fBGxv2+5q81/cPd5jm6dIsirJWJKxiYmJmUaXJE0yyGmihXR+qz8SeCnwAuCUHlNr9yZTrJtq/OmDVWurallVLRsZGZl+aElST4OcJnoj8N2qmqiq/wW+CvwWcEg7bQSwGHiw3R8HlgC09S8CdnaP99hGkjQLBimD+4Hjkjy/nfs/AbgLuB54U5uzAriy3d/Ylmnrr6uqauNntauNjgSWAt8eIJckaZoW7HlKb1W1OckVwHeAXcAtwFrgKuCyJB9rY+vaJuuALyTZRueI4Kz2OHcmuZxOkewCzq6qn840lyRp+mZcBgBVtRpYPWn4XnpcDVRVPwLOnOJxLgAuGCSLJGnmfAeyJMkykCRZBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkCctAkoRlIEnCMpAkYRlIkrAMJElYBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJIYsAySHJLkiiR3J9mS5DeTHJpkU5Kt7XZhm5skFyfZluS2JEd3Pc6KNn9rkhWD/qMkSdMz6JHBp4CvV9WvAL8GbAHOBa6tqqXAtW0Z4BRgaftaBVwCkORQYDVwLHAMsHp3gUiSZseMyyDJwcDvAOsAquonVfUosBzY0KZtAM5o95cDl1bHjcAhSY4ATgI2VdXOqnoE2AScPNNckqTpG+TI4JeACeDvk9yS5HNJXgC8pKp2ALTbw9v8RcD2ru3H29hU40+TZFWSsSRjExMTA0SXJHUbpAwWAEcDl1TVa4H/4WenhHpJj7HqM/70waq1VbWsqpaNjIxMN68kaQqDlME4MF5Vm9vyFXTK4aF2+od2+3DX/CVd2y8GHuwzLkmaJTMug6r6PrA9ySva0AnAXcBGYPcVQSuAK9v9jcBb21VFxwGPtdNI1wAnJlnYXjg+sY1JkmbJggG3fyfwxSQHAvcCb6NTMJcnWQncD5zZ5l4NnApsA55oc6mqnUnOB25q8z5aVTsHzCVJmoaByqCqbgWW9Vh1Qo+5BZw9xeOsB9YPkkWSNHO+A1mSZBlIkiwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkCctAkoRlIEli8L9noHli9NyrhvK89605bSjPK2l6PDKQJFkGkiTLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkiX1QBkkOSHJLkn9qy0cm2Zxka5IvJzmwjR/Ulre19aNdj3FeG78nyUmDZpIkTc++ODJ4N7Cla/lC4KKqWgo8Aqxs4yuBR6rqZcBFbR5JjgLOAl4FnAx8JskB+yCXJGkvDVQGSRYDpwGfa8sBjgeuaFM2AGe0+8vbMm39CW3+cuCyqvpxVX0X2AYcM0guSdL0DHpk8LfA+4En2/KLgUeraldbHgcWtfuLgO0Abf1jbf7/j/fY5imSrEoylmRsYmJiwOiSpN1mXAZJ/gB4uKpu7h7uMbX2sK7fNk8drFpbVcuqatnIyMi08kqSprZggG1fB5ye5FTgucDBdI4UDkmyoP32vxh4sM0fB5YA40kWAC8CdnaN79a9jSRpFsz4yKCqzquqxVU1SucF4Ouq6i3A9cCb2rQVwJXt/sa2TFt/XVVVGz+rXW10JLAU+PZMc0mSpm+QI4OpfAC4LMnHgFuAdW18HfCFJNvoHBGcBVBVdya5HLgL2AWcXVU/fQZySZKmsE/KoKpuAG5o9++lx9VAVfUj4Mwptr8AuGBfZJEkTZ/vQJYkWQaSJMtAkoRlIEnCMpAkYRlIkrAMJElYBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkCctAkoRlIEnCMpAkYRlIkrAMJElYBpIkLANJEgOUQZIlSa5PsiXJnUne3cYPTbIpydZ2u7CNJ8nFSbYluS3J0V2PtaLN35pkxeD/LEnSdAxyZLAL+IuqeiVwHHB2kqOAc4Frq2opcG1bBjgFWNq+VgGXQKc8gNXAscAxwOrdBSJJmh0zLoOq2lFV32n3fwhsARYBy4ENbdoG4Ix2fzlwaXXcCByS5AjgJGBTVe2sqkeATcDJM80lSZq+ffKaQZJR4LXAZuAlVbUDOoUBHN6mLQK2d2023samGu/1PKuSjCUZm5iY2BfRJUnsgzJI8kLgK8B7qurxflN7jFWf8acPVq2tqmVVtWxkZGT6YSVJPQ1UBkmeQ6cIvlhVX23DD7XTP7Tbh9v4OLCka/PFwIN9xiVJs2SQq4kCrAO2VNUnu1ZtBHZfEbQCuLJr/K3tqqLjgMfaaaRrgBOTLGwvHJ/YxiRJs2TBANu+Dvhj4PYkt7axDwJrgMuTrATuB85s664GTgW2AU8AbwOoqp1JzgduavM+WlU7B8glSZqmGZdBVf0bvc/3A5zQY34BZ0/xWOuB9TPNIkkazCBHBpqm0XOvGnYESerJj6OQJFkGkiTLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiT84zbSPjesP2J035rThvK82j94ZCBJsgwkSZaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSSJZ+lnEw3rs2Ok/ZWfxzT/eWQgSbIMJEmWgSSJOfSaQZKTgU8BBwCfq6o1Q44kaY4b5ut/+9vrFXOiDJIcAPwd8PvAOHBTko1Vdddwk0lSb/vbi+Zz5TTRMcC2qrq3qn4CXAYsH3ImSXrWmBNHBsAiYHvX8jhw7ORJSVYBq9rifye5Z4bPdxjwgxluO9vmU1aYlDcXDjHJns3rfTvZHNvX+9W+nUty4UBZf3GqFXOlDNJjrJ42ULUWWDvwkyVjVbVs0MeZDfMpK8yvvPMpK8yvvPMpK8yvvM9U1rlymmgcWNK1vBh4cEhZJOlZZ66UwU3A0iRHJjkQOAvYOORMkvSsMSdOE1XVriTnANfQubR0fVXd+Qw+5cCnmmbRfMoK8yvvfMoK8yvvfMoK8yvvM5I1VU87NS9JepaZK6eJJElDZBlIkvbvMkiyJMn1SbYkuTPJu9v4mW35ySRz5nKyPnk/nuTuJLcl+Yckh8zhrOe3nLcm+UaSlw47K0ydt2v9+5JUksOGlbEry1T79iNJHmj79tYkpw47K/Tft0nemeSeNv7Xw8zZ8ky1b7/ctV/vS3LrsLNC37yvSXJjyzuW5JiBn6yq9tsv4Ajg6Hb/54H/AI4CXgm8ArgBWDbsnHuR90RgQRu/ELhwDmc9uGvOu4DPDjtrv7xteQmdixe+Bxw2V7MCHwHeN+x808j7e8A3gYPausPnatZJc/4G+PCws+5h334DOKWNnwrcMOhz7ddHBlW1o6q+0+7/ENgCLKqqLVU103cvP2P65P1GVe1q026k8z6MoeqT9fGuaS+gx5sHh2GqvG31RcD7mR9Z55w+ed8BrKmqH7d1Dw8vZcee9m2SAH8EfGk4CZ+qT94CDm7TXsQ+eF/Wfl0G3ZKMAq8FNg83yd7pk/dPgK/Ndp5+JmdNckGS7cBbgA8PL1lv3XmTnA48UFX/PtRQU+jx/+CcdhpufZKFQws2hUl5Xw68PsnmJP+c5DeGmW2yKb7HXg88VFVbh5Gpn0l53wN8vH2ffQI4b9DHf1aUQZIXAl8B3jPpN9c5aaq8ST4E7AK+OKxsk/XKWlUfqqoldHKeM8x8k3XnpbMvP8QcLCzouW8vAX4ZeA2wg87pjDmjR94FwELgOOAvgcvbb95D1+dnwpuZI0cF3XrkfQfw3vZ99l5g3cBPMuxzYrNwzu05dM4H/3mPdTcwh14z6JcXWAF8C3j+sDPuzb5t638RuGPYOafKC/wq8DBwX/vaBdwP/MJcy9pj/ehc3rdt7OvAG7qW/xMYmYtZ2/gC4CFg8bAz7sW+fYyfvU8swOODPs9+fWTQfgtZB2ypqk8OO8+eTJU3nT/88wHg9Kp6Ylj5uvXJurRr2unA3bOdrZdeeavq9qo6vKpGq2qUzmdkHV1V3x9i1H779oiuaX8I3DHb2Xrp8332j8Dxbc7LgQMZ8ieD7uFnwhuBu6tqfPaT9dYn74PA77b7xwMDn9bar9+BnOS3gX8FbgeebMMfBA4CPg2MAI8Ct1bVSUMJ2aVP3ovpZP6vNnZjVb199hP+TJ+sK+lcqfUknatz3l5VDwwlZJep8lbV1V1z7qNzpDjsH1hT7ds30zlFVHSOZP60qnYMI2O3Pnm/Caynk/kndK6Eum4oIZt+/w+SfJ7O99Znh5Vvsj779nE6fxlyAfAj4M+q6uaBnmt/LgNJ0t7Zr08TSZL2jmUgSbIMJEmWgSQJy0CShGUgScIykCQB/wew9flEIVwuRwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(activity_score_df['sgRNA length (including PAM)'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nucleotides at each position.\n",
    "# Featurized as numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>24</th>\n",
       "      <th>25</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   1  2  3  4  5  6  7  8  9 10  ... 17 18 19 20 21 22    23    24    25  \\\n",
       "0  G  C  G  C  T  C  T  G  A  T  ...  G  A  G  C  G     None  None  None   \n",
       "1  G  C  C  C  C  A  G  G  A  T  ...  C  C  G  C  G     None  None  None   \n",
       "2  G  G  C  C  G  C  C  C  T  C  ...  C  T  C  T  G     None  None  None   \n",
       "3  G  A  C  G  G  C  G  A  C  C  ...  G  A  G  G  T     None  None  None   \n",
       "4  G  G  T  G  C  A  G  C  G  G  ...  G  G  C  G  G     None  None  None   \n",
       "\n",
       "  Length  \n",
       "0     21  \n",
       "1     21  \n",
       "2     21  \n",
       "3     21  \n",
       "4     21  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_len = max(activity_score_df['sgRNA sequence'].str.len())\n",
    "sgRNA_seq = activity_score_df['sgRNA sequence'].str.split('', expand=True)\n",
    "sgRNA_seq['Length'] = activity_score_df['sgRNA sequence'].str.len()\n",
    "# drop blank columns containing all empty strings\n",
    "sgRNA_seq.drop(columns = [0, 26],  inplace=True)\n",
    "# replace empty strings with 'None'\n",
    "sgRNA_seq.replace(to_replace='', value=None, inplace=True)\n",
    "sgRNA_seq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>24</th>\n",
       "      <th>25</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   1  2  3  4  5  6  7  8  9 10  ... 17 18 19 20 21 22    23    24    25  \\\n",
       "0  2  1  2  1  3  1  3  2  0  3  ...  2  0  2  1  2     None  None  None   \n",
       "1  2  1  1  1  1  0  2  2  0  3  ...  1  1  2  1  2     None  None  None   \n",
       "2  2  2  1  1  2  1  1  1  3  1  ...  1  3  1  3  2     None  None  None   \n",
       "3  2  0  1  2  2  1  2  0  1  1  ...  2  0  2  2  3     None  None  None   \n",
       "4  2  2  3  2  1  0  2  1  2  2  ...  2  2  1  2  2     None  None  None   \n",
       "\n",
       "  Length  \n",
       "0     21  \n",
       "1     21  \n",
       "2     21  \n",
       "3     21  \n",
       "4     21  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Need numerical fetaures for ML. \n",
    "# Nucleotides at each position.\n",
    "sgRNA_seq.replace(to_replace='A', value=0, inplace=True)\n",
    "sgRNA_seq.replace(to_replace='C', value=1, inplace=True)\n",
    "sgRNA_seq.replace(to_replace='G', value=2, inplace=True)\n",
    "sgRNA_seq.replace(to_replace='T', value=3, inplace=True)\n",
    "sgRNA_seq.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overall nucleotide composition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>24</th>\n",
       "      <th>25</th>\n",
       "      <th>Length</th>\n",
       "      <th>As</th>\n",
       "      <th>Cs</th>\n",
       "      <th>Gs</th>\n",
       "      <th>Ts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.238095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.523810</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.095238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.190476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.238095</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.095238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.523810</td>\n",
       "      <td>0.190476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   1  2  3  4  5  6  7  8  9 10  ... 21 22    23    24    25 Length        As  \\\n",
       "0  2  1  2  1  3  1  3  2  0  3  ...  2     None  None  None     21  0.142857   \n",
       "1  2  1  1  1  1  0  2  2  0  3  ...  2     None  None  None     21  0.142857   \n",
       "2  2  2  1  1  2  1  1  1  3  1  ...  2     None  None  None     21  0.095238   \n",
       "3  2  0  1  2  2  1  2  0  1  1  ...  3     None  None  None     21  0.238095   \n",
       "4  2  2  3  2  1  0  2  1  2  2  ...  2     None  None  None     21  0.047619   \n",
       "\n",
       "         Cs        Gs        Ts  \n",
       "0  0.238095  0.428571  0.238095  \n",
       "1  0.523810  0.333333  0.095238  \n",
       "2  0.380952  0.428571  0.190476  \n",
       "3  0.238095  0.476190  0.095238  \n",
       "4  0.333333  0.523810  0.190476  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "As = []\n",
    "Cs = []\n",
    "Gs = []\n",
    "Ts = []\n",
    "for index, row in sgRNA_seq.iterrows():\n",
    "    As.append(sum(row == 0)/row['Length'])\n",
    "    Cs.append(sum(row == 1)/row['Length'])\n",
    "    Gs.append(sum(row == 2)/row['Length'])\n",
    "    Ts.append(sum(row == 3)/row['Length'])\n",
    "sgRNA_seq['As'] = As\n",
    "sgRNA_seq['Cs'] = Cs\n",
    "sgRNA_seq['Gs'] = Gs\n",
    "sgRNA_seq['Ts'] = Ts\n",
    "sgRNA_seq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([2.516e+03, 3.359e+03, 5.327e+03, 4.537e+03, 1.651e+03, 7.380e+02,\n",
       "        1.760e+02, 6.100e+01, 1.000e+01, 5.000e+00]),\n",
       " array([0.        , 0.07368421, 0.14736842, 0.22105263, 0.29473684,\n",
       "        0.36842105, 0.44210526, 0.51578947, 0.58947368, 0.66315789,\n",
       "        0.73684211]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD4CAYAAADlwTGnAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAARxklEQVR4nO3df6zddX3H8edL6o/NKSBUw9qy4qwTXP21DklMFhUHlW2WZWBq5qymWzPHNpe5TJhLYKhRt0Smmbp1Qqxmih2bo1PRVX7EaIZSBalQkYpMGghUCzhHZKu+98f5FA7ltvfc23vPPeXzfCQ35/t9fz/nfN/n2977ut9f56aqkCT153EL3YAkaWEYAJLUKQNAkjplAEhSpwwASerUooVu4GCOPfbYWr58+UK3IUmHla9+9avfq6rF042b6ABYvnw527ZtW+g2JOmwkuS/RhnnISBJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASerURN8JrNHseM6JY1/nid/cMfZ1Sppb7gFIUqfcA9Chu+DIBVrv/QuzXukxwj0ASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1KmRAiDJ7Um2J7khybZWe1qSrUlubY9Ht3qSvC/JziQ3JnnR0Ousa+NvTbJuft6SJGkUM9kDeFlVvaCqVrX5c4Erq2oFcGWbB3glsKJ9bQA+CIPAAM4HXgycDJy/LzQkSeN3KDeCrQFe2qY3AdcAb2n1j1RVAdcmOSrJcW3s1qraA5BkK7Aa+Pgh9CCNZOWmlWNf5/Z128e+TmkmRg2AAv4jSQH/UFUbgWdU1V0AVXVXkqe3sUuAO4aeu6vVDlR/hCQbGOw5cPzxx8/grRw+lp/76Tl9vSvm9NUk9WLUAHhJVd3ZfshvTfLNg4zNFLU6SP2RhUG4bARYtWrVo5ZLkubGSOcAqurO9ngP8EkGx/Dvbod2aI/3tOG7gGVDT18K3HmQuiRpAUwbAEmenOQp+6aB04BvAFuAfVfyrAMub9NbgNe1q4FOAe5vh4o+B5yW5Oh28ve0VpMkLYBRDgE9A/hkkn3jP1ZVn01yHbA5yXrgu8DZbfxngDOAncADwBsAqmpPkrcB17VxF+47ISxJGr9pA6CqbgOeP0X9+8CpU9QLOOcAr3UJcMnM25QkzTXvBJakThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpw7lT0KqYzuec+LQ3M+OZZ0nrvXPR0hzyT0ASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlHcCH8Tycz+90C1I0rxxD0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ystApXmyctPKsa5v+7rtY12fDn8j7wEkOSLJ9Uk+1eZPSPLlJLcm+USSJ7T6E9v8zrZ8+dBrnNfqtyQ5fa7fjCRpdDM5BPQmYMfQ/LuBi6pqBXAvsL7V1wP3VtWzgIvaOJKcBKwFngusBj6Q5IhDa1+SNFsjBUCSpcCvAR9q8wFeDlzWhmwCzmzTa9o8bfmpbfwa4NKqerCqvgPsBE6eizchSZq5UfcA/hb4c+Anbf4Y4L6q2tvmdwFL2vQS4A6Atvz+Nv6h+hTPeUiSDUm2Jdm2e/fuGbwVSdJMTBsASX4duKeqvjpcnmJoTbPsYM95uFC1sapWVdWqxYsXT9eeJGmWRrkK6CXAq5KcATwJeCqDPYKjkixqv+UvBe5s43cBy4BdSRYBRwJ7hur7DD9HkjRm0+4BVNV5VbW0qpYzOIl7VVX9NnA1cFYbtg64vE1vafO05VdVVbX62naV0AnACuArc/ZOJEkzcij3AbwFuDTJ24HrgYtb/WLgo0l2MvjNfy1AVd2UZDNwM7AXOKeqfnwI65ckHYIZBUBVXQNc06ZvY4qreKrqR8DZB3j+O4B3zLRJSdLc86MgJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4dyh+F1xSu+Lc/W+gWJGkk7gFIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6tS0AZDkSUm+kuTrSW5K8letfkKSLye5Ncknkjyh1Z/Y5ne25cuHXuu8Vr8lyenz9aYkSdMbZQ/gQeDlVfV84AXA6iSnAO8GLqqqFcC9wPo2fj1wb1U9C7iojSPJScBa4LnAauADSY6YyzcjSRrdtAFQAz9ss49vXwW8HLis1TcBZ7bpNW2etvzUJGn1S6vqwar6DrATOHlO3oUkacZGOgeQ5IgkNwD3AFuBbwP3VdXeNmQXsKRNLwHuAGjL7weOGa5P8ZzhdW1Isi3Jtt27d8/8HUmSRjLSx0FX1Y+BFyQ5CvgkcOJUw9pjDrDsQPX917UR2AiwatWqRy2XHnLBkaOPPeH4+etDOkzN6CqgqroPuAY4BTgqyb4AWQrc2aZ3AcsA2vIjgT3D9SmeI0kas1GuAlrcfvMnyU8BrwB2AFcDZ7Vh64DL2/SWNk9bflVVVauvbVcJnQCsAL4yV29EkjQzoxwCOg7Y1K7YeRywuao+leRm4NIkbweuBy5u4y8GPppkJ4Pf/NcCVNVNSTYDNwN7gXPaoSVJ0gKYNgCq6kbghVPUb2OKq3iq6kfA2Qd4rXcA75h5m5KkueadwJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqdG+nsA0iRY6Wf6S3PKPQBJ6tRjeg9gx3Om+sNlo7tijvqQpEnkHoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVPTBkCSZUmuTrIjyU1J3tTqT0uyNcmt7fHoVk+S9yXZmeTGJC8aeq11bfytSdbN39uSJE1nlD2AvcCbq+pE4BTgnCQnAecCV1bVCuDKNg/wSmBF+9oAfBAGgQGcD7wYOBk4f19oSJLGb9oAqKq7quprbfq/gR3AEmANsKkN2wSc2abXAB+pgWuBo5IcB5wObK2qPVV1L7AVWD2n70aSNLIZnQNIshx4IfBl4BlVdRcMQgJ4ehu2BLhj6Gm7Wu1A9f3XsSHJtiTbdu/ePZP2JEkzMHIAJPkZ4F+AP6mqHxxs6BS1Okj9kYWqjVW1qqpWLV68eNT2JEkzNFIAJHk8gx/+/1RV/9rKd7dDO7THe1p9F7Bs6OlLgTsPUpckLYBRrgIKcDGwo6reM7RoC7DvSp51wOVD9de1q4FOAe5vh4g+B5yW5Oh28ve0VpMkLYBFI4x5CfA7wPYkN7TaXwDvAjYnWQ98Fzi7LfsMcAawE3gAeANAVe1J8jbgujbuwqraMyfvQhIrN60c+zq3r9s+9nVq7kwbAFX1RaY+fg9w6hTjCzjnAK91CXDJTBqUJM0P7wSWpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdWqUO4GlibD5nXvHvs5Xn+e3iB673AOQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6tS0AZDkkiT3JPnGUO1pSbYmubU9Ht3qSfK+JDuT3JjkRUPPWdfG35pk3fy8HUnSqEbZA/gwsHq/2rnAlVW1AriyzQO8EljRvjYAH4RBYADnAy8GTgbO3xcakqSFMW0AVNUXgD37ldcAm9r0JuDMofpHauBa4KgkxwGnA1urak9V3Qts5dGhIkkao9meA3hGVd0F0B6f3upLgDuGxu1qtQPVHyXJhiTbkmzbvXv3LNuTJE1nrk8CZ4paHaT+6GLVxqpaVVWrFi9ePKfNSZIetmiWz7s7yXFVdVc7xHNPq+8Clg2NWwrc2eov3a9+zSzXLWlCrNy0cuzr3L5u+9jX+Vg12z2ALcC+K3nWAZcP1V/XrgY6Bbi/HSL6HHBakqPbyd/TWk2StECm3QNI8nEGv70fm2QXg6t53gVsTrIe+C5wdhv+GeAMYCfwAPAGgKrak+RtwHVt3IVVtf+JZUnSGE0bAFX1mgMsOnWKsQWcc4DXuQS4ZEbdSZLmjXcCS1KnDABJ6pQBIEmdMgAkqVOzvQ9A6sLmd+4d+zpffZ7flhoP9wAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOrVooRuQ9Eib37l37Ot89XmHz4+ClZtWjn2d29dtH/s6x8E9AEnqlAEgSZ0yACSpU2M/8JdkNfBe4AjgQ1X1rnH3IOmRPO/Qp7H+CyQ5Ang/8KvALuC6JFuq6uZx9iFJM/FYPfE87gg+GdhZVbcBJLkUWAMYAFJn3OtYeOPeGkuAO4bmdwEvHh6QZAOwoc3+MMkth7C+Y4HvHcLz55O9zd4k92dvszf//b1+1s8c+7bL6zOT4fv393OjPGncATDVO6pHzFRtBDbOycqSbVW1ai5ea67Z2+xNcn/2NnuT3N8k9waz72/cVwHtApYNzS8F7hxzD5Ikxh8A1wErkpyQ5AnAWmDLmHuQJDHmQ0BVtTfJHwKfY3AZ6CVVddM8rnJODiXNE3ubvUnuz95mb5L7m+TeYJb9paqmHyVJeszxTmBJ6pQBIEmdOuwDIMnqJLck2Znk3CmWPzHJJ9ryLydZPmH9/UqSryXZm+SsCevtT5PcnOTGJFcmGena4jH19vtJtie5IckXk5w0rt5G6W9o3FlJKsnYLiEcYdu9Psnutu1uSPK74+ptlP7amFe3/3s3JfnYpPSW5KKh7fatJPdNUG/HJ7k6yfXte/aMaV+0qg7bLwYnkr8NPBN4AvB14KT9xvwB8Pdtei3wiQnrbznwPOAjwFkT1tvLgJ9u028c17YbsbenDk2/CvjsJG27Nu4pwBeAa4FVk9Ibg9uh/m5c22sW/a0ArgeObvNPn5Te9hv/RwwuZJmI3hicCH5jmz4JuH261z3c9wAe+miJqvpfYN9HSwxbA2xq05cBpyaZ0S1289lfVd1eVTcCPxlTTzPp7eqqeqDNXsvgvo1J6e0HQ7NPZr8bChe6v+ZtwF8DP5rA3hbKKP39HvD+qroXoKrumaDehr0G+PhYOhuttwKe2qaPZIR7rA73AJjqoyWWHGhMVe0F7geOGUt3o/W3UGba23rginnt6GEj9ZbknCTfZvBD9o/H1BuM0F+SFwLLqupTY+wLRv93/a12mOCyJMumWD5fRunv2cCzk3wpybXtE4QnpTcA2uHQE4CrxtAXjNbbBcBrk+wCPsNgD+WgDvcAmPajJUYcM18Wct3TGbm3JK8FVgF/M68dDa1yitqjequq91fVzwNvAf5y3rt62EH7S/I44CLgzWPr6GGjbLt/B5ZX1fOAz/PwHvI4jNLfIgaHgV7K4LfsDyU5ap77gpl9v64FLquqH89jP8NG6e01wIerailwBvDR9n/xgA73ABjloyUeGpNkEYNdoz1j6W6yP/pipN6SvAJ4K/Cqqnpwknobcilw5rx29EjT9fcU4BeBa5LcDpwCbBnTieBpt11VfX/o3/IfgV8aQ1/7jPo9e3lV/V9VfQe4hUEgTEJv+6xlfId/YLTe1gObAarqP4EnMfiQuAMbxwmMeTwxsgi4jcGu2L4TI8/db8w5PPIk8OZJ6m9o7IcZ70ngUbbdCxmceFoxgf+uK4amfwPYNkn97Tf+GsZ3EniUbXfc0PRvAtdO0rYDVgOb2vSxDA59HDMJvbVxvwDcTruRdoK22xXA69v0iQwC4qA9jqX5ed4wZwDfaj+o3tpqFzL4jRUGKfjPwE7gK8AzJ6y/X2aQ7v8DfB+4aYJ6+zxwN3BD+9oyQb29F7ip9XX1wX4AL0R/+40dWwCMuO3e2bbd19u2e84kbTsGhzvew+DvhGwH1k5Kb23+AuBd49xmI263k4AvtX/XG4DTpntNPwpCkjp1uJ8DkCTNkgEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOvX//xVx8u8LjrUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Look at nucleotide distribution\n",
    "plt.hist(sgRNA_seq['As'])\n",
    "plt.hist(sgRNA_seq['Cs'])\n",
    "plt.hist(sgRNA_seq['Gs'])\n",
    "plt.hist(sgRNA_seq['Ts'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Nucleotide Dimers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "def break_seq(seq):\n",
    "    return [char for char in seq]\n",
    "\n",
    "def break_seq_dimer(seq):\n",
    "    return [seq[i:i+2] for i in range(0, len(seq)-1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18375</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18376</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18377</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18378</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18379</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18380 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  1  2  3  4  5  6  7  8  9   ... 15 16 17 18 19    20    21    22  \\\n",
       "0      G  C  G  C  T  C  T  G  A  T  ...  G  G  A  G  C     G  None  None   \n",
       "1      G  C  C  C  C  A  G  G  A  T  ...  C  C  C  G  C     G  None  None   \n",
       "2      G  G  C  C  G  C  C  C  T  C  ...  G  C  T  C  T     G  None  None   \n",
       "3      G  A  C  G  G  C  G  A  C  C  ...  A  G  A  G  G     T  None  None   \n",
       "4      G  G  T  G  C  A  G  C  G  G  ...  T  G  G  C  G     G  None  None   \n",
       "...   .. .. .. .. .. .. .. .. .. ..  ... .. .. .. .. ..   ...   ...   ...   \n",
       "18375  G  T  A  A  T  C  T  T  A  G  ...  G  T  G  G  G     A  None  None   \n",
       "18376  G  T  G  A  C  T  T  C  A  A  ...  T  C  T  T  A  None  None  None   \n",
       "18377  G  T  G  G  A  T  G  T  G  G  ...  G  G  C  G  A     A  None  None   \n",
       "18378  G  A  C  A  C  C  T  C  A  G  ...  T  A  A  G  A     C  None  None   \n",
       "18379  G  G  G  C  A  C  C  A  A  T  ...  C  A  C  A  T  None  None  None   \n",
       "\n",
       "         23    24  \n",
       "0      None  None  \n",
       "1      None  None  \n",
       "2      None  None  \n",
       "3      None  None  \n",
       "4      None  None  \n",
       "...     ...   ...  \n",
       "18375  None  None  \n",
       "18376  None  None  \n",
       "18377  None  None  \n",
       "18378  None  None  \n",
       "18379  None  None  \n",
       "\n",
       "[18380 rows x 25 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pd.DataFrame(activity_score_df['sgRNA sequence'].apply(break_seq))\n",
    "pd.DataFrame(activity_score_df['sgRNA sequence'].apply(break_seq).tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   0   1   2   3   4  5   6  7   8   9   ...  14  15  16  17  18  19    20  \\\n",
       "0   9   6   9   7  13  7  14  8   3  15  ...   6  10   8   2   9   6  None   \n",
       "1   9   5   5   5   4  2  10  8   3  13  ...   5   5   5   6   9   6  None   \n",
       "2  10   9   5   6   9  5   5  7  13   6  ...   2   9   7  13   7  14  None   \n",
       "3   8   1   6  10   9  6   8  1   5   5  ...   8   2   8   2  10  11  None   \n",
       "4  10  11  14   9   4  2   9  6  10  10  ...  15  14  10   9   6  10  None   \n",
       "\n",
       "     21    22    23  \n",
       "0  None  None  None  \n",
       "1  None  None  None  \n",
       "2  None  None  None  \n",
       "3  None  None  None  \n",
       "4  None  None  None  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Nucleotide dimers\n",
    "encodings = {'AA' : 0, 'AC' : 1, 'AG' : 2, 'AT' : 3, 'CA' : 4, 'CC' : 5, 'CG' : 6, 'CT' : 7, \n",
    "             'GA' : 8, 'GC' : 9, 'GG' : 10, 'GT' : 11, 'TA' : 12, 'TC' : 13, 'TG' : 14, 'TT' : 15}\n",
    "sgRNA_dimers = pd.DataFrame(activity_score_df['sgRNA sequence'].apply(break_seq_dimer).tolist())\n",
    "\n",
    "for dimer, repl in encodings.items():\n",
    "    sgRNA_dimers.replace(dimer, repl, inplace = True)\n",
    "    \n",
    "sgRNA_dimers.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overall dimer composition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>GC</th>\n",
       "      <th>CA</th>\n",
       "      <th>AG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    0   1   2   3   4  5   6  7   8   9  ...  17  18  19    20    21    22  \\\n",
       "0   9   6   9   7  13  7  14  8   3  15  ...   2   9   6  None  None  None   \n",
       "1   9   5   5   5   4  2  10  8   3  13  ...   6   9   6  None  None  None   \n",
       "2  10   9   5   6   9  5   5  7  13   6  ...  13   7  14  None  None  None   \n",
       "3   8   1   6  10   9  6   8  1   5   5  ...   2  10  11  None  None  None   \n",
       "4  10  11  14   9   4  2   9  6  10  10  ...   9   6  10  None  None  None   \n",
       "\n",
       "     23 GC CA  AG  \n",
       "0  None  0  0   9  \n",
       "1  None  0  0   7  \n",
       "2  None  0  0   8  \n",
       "3  None  0  0   9  \n",
       "4  None  0  0  11  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Relevant dimers that the paper decsribes: GC, CA, AG\n",
    "GC = []\n",
    "CA = []\n",
    "AG = []\n",
    "for index, row in sgRNA_seq.iterrows():\n",
    "    GC.append(sum(row == encodings['GC']))\n",
    "    CA.append(sum(row == encodings['CA']))\n",
    "    AG.append(sum(row == encodings['AG']))\n",
    "\n",
    "sgRNA_dimers['GC'] = GC\n",
    "sgRNA_dimers['CA'] = CA\n",
    "sgRNA_dimers['AG'] = AG\n",
    "              \n",
    "sgRNA_dimers.head()             "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([ 715.,  615., 2982., 2560., 5830., 2419., 2607.,  441.,  199.,\n",
       "          12.]),\n",
       " array([ 1. ,  2.5,  4. ,  5.5,  7. ,  8.5, 10. , 11.5, 13. , 14.5, 16. ]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAD4CAYAAAAO9oqkAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAUV0lEQVR4nO3df4xd9Znf8fenJtD9kQQTG+pgWEPkRIKo65ARoU0T0VLAoFVMqk1rtFrcBMkhBWmjtlKcRio0W6Rkd7ORqFIiZ2PFVFl+bAiLlTUlXjZaVCkQD8QBO4R4ICQMdm0HU2DFiq3p0z/ud7In4zv2eO74zgW/X9LVPfc533PvM9dH8/H5nnPvpKqQJJ3Y/sFCNyBJWniGgSTJMJAkGQaSJAwDSRJw0kI3MFdLliypFStWLHQbkvS68sgjj/y8qpZOr79uw2DFihWMj48vdBuS9LqS5Kf96k4TSZIMA0mSYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw2BGKzb8BSs2/MVCtyFJQ2EYSJIMA0mSYSBJwjCQJGEYSJKYRRgk2ZRkf5KdndqdSXa02zNJdrT6iiR/21n35c42703yeJKJJLckSauflmRbkt3tfvHx+EElSTObzZHB14DV3UJV/ZuqWlVVq4C7gW92Vj81ta6qruvUbwXWAyvbbeo5NwAPVNVK4IH2WJI0REcNg6p6EDjYb1373/2/Bm4/0nMkWQa8paq+W1UF3AZc1VavATa35c2duiRpSAY9Z/ABYF9V7e7Uzkny/SR/neQDrXYmMNkZM9lqAGdU1V6Adn/6gD1Jko7RoH8D+Wp++ahgL3B2VT2f5L3Anyc5H0ifbetYXyzJenpTTZx99tlzaFeS1M+cjwySnAT8K+DOqVpVvVpVz7flR4CngHfSOxJY3tl8ObCnLe9r00hT00n7Z3rNqtpYVWNVNbZ06dK5ti5JmmaQaaJ/Cfyoqn4x/ZNkaZJFbflceieKn27TPy8nuaidZ7gGuLdttgVY15bXdeqSpCGZzaWltwPfBd6VZDLJtW3VWg4/cfxB4LEkPwC+AVxXVVMnnz8B/AkwQe+I4b5W/xxwaZLdwKXtsSRpiI56zqCqrp6h/m/71O6md6lpv/HjwLv71J8HLjlaH5Kk48dPIEuSDANJkmEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJAnDQJKEYSBJwjCQJDGLMEiyKcn+JDs7tZuSPJdkR7td2Vn36SQTSZ5McnmnvrrVJpJs6NTPSfJwkt1J7kxy8nz+gJKko5vNkcHXgNV96l+sqlXtthUgyXnAWuD8ts1/T7IoySLgS8AVwHnA1W0swOfbc60EXgCuHeQHkiQdu6OGQVU9CByc5fOtAe6oqler6ifABHBhu01U1dNV9XfAHcCaJAH+BfCNtv1m4Kpj/BkkSQMa5JzBDUkea9NIi1vtTODZzpjJVpup/jbg/1TVoWn1vpKsTzKeZPzAgQMDtC5J6pprGNwKvANYBewFvtDq6TO25lDvq6o2VtVYVY0tXbr02DqWJM3opLlsVFX7ppaTfAX4Vns4CZzVGboc2NOW+9V/Dpya5KR2dNAdL0kakjkdGSRZ1nn4YWDqSqMtwNokpyQ5B1gJfA/YDqxsVw6dTO8k85aqKuA7wG+37dcB986lJ0nS3B31yCDJ7cDFwJIkk8CNwMVJVtGb0nkG+DhAVe1KchfwQ+AQcH1Vvdae5wbgfmARsKmqdrWX+BRwR5L/Cnwf+Oq8/XSSpFk5ahhU1dV9yjP+wq6qm4Gb+9S3Alv71J+md7WRJGmB+AlkSZJhIEkyDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEnMIgySbEqyP8nOTu0Pk/woyWNJ7klyaquvSPK3SXa025c727w3yeNJJpLckiStflqSbUl2t/vFx+MHlSTNbDZHBl8DVk+rbQPeXVX/GPgx8OnOuqeqalW7Xdep3wqsB1a229RzbgAeqKqVwAPtsSRpiI4aBlX1IHBwWu3bVXWoPXwIWH6k50iyDHhLVX23qgq4DbiqrV4DbG7Lmzt1SdKQzMc5g48B93Uen5Pk+0n+OskHWu1MYLIzZrLVAM6oqr0A7f70mV4oyfok40nGDxw4MA+tS5JgwDBI8hngEPD1VtoLnF1V7wH+PfCnSd4CpM/mdayvV1Ubq2qsqsaWLl0617YlSdOcNNcNk6wDfgu4pE39UFWvAq+25UeSPAW8k96RQHcqaTmwpy3vS7Ksqva26aT9c+1JkjQ3czoySLIa+BTwoap6pVNfmmRRWz6X3onip9v0z8tJLmpXEV0D3Ns22wKsa8vrOnVJ0pAc9cggye3AxcCSJJPAjfSuHjoF2NauEH2oXTn0QeCzSQ4BrwHXVdXUyedP0Lsy6VfonWOYOs/wOeCuJNcCPwM+Mi8/mSRp1o4aBlV1dZ/yV2cYezdw9wzrxoF396k/D1xytD4kScePn0CWJBkGkiTDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAksQswyDJpiT7k+zs1E5Lsi3J7na/uNWT5JYkE0keS3JBZ5t1bfzuJOs69fcmebxtc0uSzOcPKUk6stkeGXwNWD2ttgF4oKpWAg+0xwBXACvbbT1wK/TCA7gReB9wIXDjVIC0Mes7201/LUnScTSrMKiqB4GD08prgM1teTNwVad+W/U8BJyaZBlwObCtqg5W1QvANmB1W/eWqvpuVRVwW+e5JElDMMg5gzOqai9Auz+91c8Enu2Mm2y1I9Un+9QPk2R9kvEk4wcOHBigdUlS1/E4gdxvvr/mUD+8WLWxqsaqamzp0qUDtChJ6hokDPa1KR7a/f5WnwTO6oxbDuw5Sn15n7okaUgGCYMtwNQVQeuAezv1a9pVRRcBL7ZppPuBy5IsbieOLwPub+teTnJRu4roms5zSZKG4KTZDEpyO3AxsCTJJL2rgj4H3JXkWuBnwEfa8K3AlcAE8ArwUYCqOpjk94Htbdxnq2rqpPQn6F2x9CvAfe0mSRqSWYVBVV09w6pL+owt4PoZnmcTsKlPfRx492x6kSTNPz+BLEkyDCRJhoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSSJAcIgybuS7OjcXkryySQ3JXmuU7+ys82nk0wkeTLJ5Z366labSLJh0B9KknRsTprrhlX1JLAKIMki4DngHuCjwBer6o+645OcB6wFzgfeDvxlkne21V8CLgUmge1JtlTVD+famyTp2Mw5DKa5BHiqqn6aZKYxa4A7qupV4CdJJoAL27qJqnoaIMkdbaxhIElDMl/nDNYCt3ce35DksSSbkixutTOBZztjJlttpvphkqxPMp5k/MCBA/PUuiRp4DBIcjLwIeDPWulW4B30ppD2Al+YGtpn8zpC/fBi1caqGquqsaVLlw7UtyTp783HNNEVwKNVtQ9g6h4gyVeAb7WHk8BZne2WA3va8kx1SdIQzMc00dV0poiSLOus+zCwsy1vAdYmOSXJOcBK4HvAdmBlknPaUcbaNlaSNCQDHRkk+VV6VwF9vFP+gySr6E31PDO1rqp2JbmL3onhQ8D1VfVae54bgPuBRcCmqto1SF+SpGMzUBhU1SvA26bVfvcI428Gbu5T3wpsHaQXSdLc+QlkSZJhIEkyDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kS8/c3kKU3npveOuTXe3G4ryd1eGQgSTIMJEmGgSQJw0CShGEgSWIewiDJM0keT7IjyXirnZZkW5Ld7X5xqyfJLUkmkjyW5ILO86xr43cnWTdoX5Kk2ZuvI4N/XlWrqmqsPd4APFBVK4EH2mOAK4CV7bYeuBV64QHcCLwPuBC4cSpAJEnH3/GaJloDbG7Lm4GrOvXbquch4NQky4DLgW1VdbCqXgC2AauPU2+SpGnmIwwK+HaSR5Ksb7UzqmovQLs/vdXPBJ7tbDvZajPVf0mS9UnGk4wfOHBgHlqXJMH8fAL5/VW1J8npwLYkPzrC2PSp1RHqv1yo2ghsBBgbGztsvSRpbgY+MqiqPe1+P3APvTn/fW36h3a/vw2fBM7qbL4c2HOEuiRpCAYKgyS/luTNU8vAZcBOYAswdUXQOuDetrwFuKZdVXQR8GKbRrofuCzJ4nbi+LJWkyQNwaDTRGcA9ySZeq4/rar/mWQ7cFeSa4GfAR9p47cCVwITwCvARwGq6mCS3we2t3GfraqDA/YmSZqlgcKgqp4GfrNP/Xngkj71Aq6f4bk2AZsG6UeSNDd+AlmSZBhIkgwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkM/mcvdaK76a1Dfr0Xh/t60gnCIwNJ0tzDIMlZSb6T5Ikku5L8XqvflOS5JDva7crONp9OMpHkySSXd+qrW20iyYbBfiRJ0rEaZJroEPAfqurRJG8GHkmyra37YlX9UXdwkvOAtcD5wNuBv0zyzrb6S8ClwCSwPcmWqvrhAL1Jko7BnMOgqvYCe9vyy0meAM48wiZrgDuq6lXgJ0kmgAvbuomqehogyR1trGEgHU/DPN/juZ6RNy8nkJOsAN4DPAy8H7ghyTXAOL2jhxfoBcVDnc0m+fvweHZa/X3z0ZfegIZ9wlo6QQx8AjnJrwN3A5+sqpeAW4F3AKvoHTl8YWpon83rCPV+r7U+yXiS8QMHDgzauiSpGejIIMmb6AXB16vqmwBVta+z/ivAt9rDSeCszubLgT1teab6L6mqjcBGgLGxsb6BIb1uedSjBTTI1UQBvgo8UVV/3Kkv6wz7MLCzLW8B1iY5Jck5wErge8B2YGWSc5KcTO8k85a59iVJOnaDHBm8H/hd4PEkO1rtPwFXJ1lFb6rnGeDjAFW1K8ld9E4MHwKur6rXAJLcANwPLAI2VdWuAfqSJB2jQa4m+l/0n+/feoRtbgZu7lPfeqTtJEnHl59AliQZBpIkw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkM+GcvNUvD/HOGN704vNeS9IZhGLzR+Hd0Jc2B00SSJI8MJA3BsI9YnS49Zh4ZSJJGJwySrE7yZJKJJBsWuh9JOpGMxDRRkkXAl4BLgUlge5ItVfXDhe1M0uuS01LHbFSODC4EJqrq6ar6O+AOYM0C9yRJJ4yRODIAzgSe7TyeBN43fVCS9cD69vBvkjx5HHtaAvw8nz+Or3B8LAF+vtBNzMHrtW94/fZu3/Plv2Q2o0al79/oVxyVMOj3TtZhhaqNwMbj3w4kGa+qsWG81nyy7+F7vfZu38M16n2PyjTRJHBW5/FyYM8C9SJJJ5xRCYPtwMok5yQ5GVgLbFngniTphDES00RVdSjJDcD9wCJgU1XtWuC2hjIddRzY9/C9Xnu37+Ea6b5TddjUvCTpBDMq00SSpAVkGEiSDIOjfQ1GklOS3NnWP5xkxfC7PKyns5J8J8kTSXYl+b0+Yy5O8mKSHe32nxei1+mSPJPk8dbTeJ/1SXJLe78fS3LBQvQ5rad3dd7HHUleSvLJaWNG5v1OsinJ/iQ7O7XTkmxLsrvdL55h23VtzO4k64bX9Yx9/2GSH7V94Z4kp86w7RH3q+Nphr5vSvJcZ3+4coZtR+dreKrqhL3RO1n9FHAucDLwA+C8aWP+HfDltrwWuHME+l4GXNCW3wz8uE/fFwPfWuhe+/T+DLDkCOuvBO6j99mTi4CHF7rnPvvM/wZ+Y1Tfb+CDwAXAzk7tD4ANbXkD8Pk+250GPN3uF7flxQvc92XASW358/36ns1+tQB93wT8x1nsS0f8/TPM24l+ZDCbr8FYA2xuy98ALkkyq48bHi9VtbeqHm3LLwNP0PsU9xvBGuC26nkIODXJsoVuquMS4Kmq+ulCNzKTqnoQODit3N2PNwNX9dn0cmBbVR2sqheAbcDq49boNP36rqpvV9Wh9vAhep9BGikzvN+zMVJfw3Oih0G/r8GY/kv1F2PaTvki8LahdDcLbdrqPcDDfVb/kyQ/SHJfkvOH2tjMCvh2kkfa14tMN5t/k4W0Frh9hnWj+H5POaOq9kLvPxPA6X3GjPp7/zF6R439HG2/Wgg3tOmtTTNMy43U+32ih8FsvgZjVl+VsRCS/DpwN/DJqnpp2upH6U1l/Cbw34A/H3Z/M3h/VV0AXAFcn+SD09aP8vt9MvAh4M/6rB7V9/tYjPJ7/xngEPD1GYYcbb8atluBdwCrgL3AF/qMGan3+0QPg9l8DcYvxiQ5CXgrczsknFdJ3kQvCL5eVd+cvr6qXqqqv2nLW4E3JVky5DYPU1V72v1+4B56h8pdo/zVJFcAj1bVvukrRvX97tg3Nd3W7vf3GTOS7307kf1bwO9Um2yfbhb71VBV1b6qeq2q/h/wlRn6Gan3+0QPg9l8DcYWYOqqit8G/mqmHXJY2jmLrwJPVNUfzzDmH02d20hyIb1/6+eH12Xfnn4tyZunlumdHNw5bdgW4Jp2VdFFwItT0xsj4GpmmCIaxfd7mu5+vA64t8+Y+4HLkixu0xqXtdqCSbIa+BTwoap6ZYYxs9mvhmraea4P07+f0foanoU6cz0qN3pXr/yY3ln9z7TaZ+ntfAD/kN60wATwPeDcEej5n9E7nHwM2NFuVwLXAde1MTcAu+hdofAQ8E9HoO9zWz8/aL1Nvd/dvkPvDx09BTwOjC10362vX6X3y/2tndpIvt/0Amsv8H/p/e/zWnrnuR4Adrf709rYMeBPOtt+rO3rE8BHR6DvCXrz6lP7+dSVfW8Hth5pv1rgvv9H238fo/cLftn0vtvjw37/LNTNr6OQJJ3w00SSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSgP8PvN5uwTd9QzYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# See distribution\n",
    "# plt.hist(sgRNA_dimers['GC'])\n",
    "# plt.hist(sgRNA_dimers['CA'])\n",
    "# plt.hist(sgRNA_dimers['AG'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for the TSS info, copy thsat over for all 20 rows from the same gene\n",
    "gene1 TSS\n",
    "gene1 TSS\n",
    "gene1 TSS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1539"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = activity_score_df['gene symbol'].unique()\n",
    "genes\n",
    "len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     A1BG\n",
       "1     A1BG\n",
       "2     A1CF\n",
       "3      A2M\n",
       "4    A2ML1\n",
       "Name: gene, dtype: object"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_tss = tss_annotation_df['gene'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "dup_genes_tss = tss_annotation_df['gene'][tss_annotation_df['gene'].duplicated()].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "75"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(np.isin(genes, dup_genes_tss))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ATP5A1', 'ATP5J2-PTCD1', 'BCL2L1', 'BOD1L1', 'C11orf45', 'CANX',\n",
       "       'CBFA2T3', 'CCNK', 'CDK6', 'CENPA', 'CKAP5', 'COPS7B', 'COPZ1',\n",
       "       'DDX46', 'DDX54', 'DHODH', 'DNASE1L1', 'DPH1', 'DYNLRB1', 'EEF2',\n",
       "       'EIF4G2', 'ESF1', 'FBXO8', 'FDFT1', 'FIS1', 'FOLR3', 'GON4L',\n",
       "       'HAUS7', 'HMG20B', 'IDH3A', 'IFT74', 'IK', 'IL11RA', 'KMT2B',\n",
       "       'LDB1', 'LSM5', 'MAPKAP1', 'MRPL30', 'MTHFD2', 'NDUFC1', 'NDUFS5',\n",
       "       'NDUFV1', 'OAZ3', 'PKM', 'PMPCB', 'PNISR', 'POLR1D', 'POLR2D',\n",
       "       'PPP2R1A', 'PRMT5', 'PSMB6', 'PSPH', 'PTCD1', 'RAB5C', 'RBM4',\n",
       "       'RPS29', 'RSL1D1', 'SEC61A1', 'SF3A1', 'SLC12A9', 'SLC1A5', 'SNW1',\n",
       "       'ST3GAL2', 'STAG2', 'TACC3', 'TAL1', 'TKT', 'TUBB', 'UBA1',\n",
       "       'USP39', 'WBP11', 'ZMAT2', 'ZMYM3', 'ZNF574', 'ZNF699'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[np.isin(genes, dup_genes_tss)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_tss = tss_annotation_df['gene'].unique()\n",
    "sum(np.logical_not(np.isin(genes, genes_tss)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NIFK'], dtype=object)"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[np.logical_not(np.isin(genes, genes_tss))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: the NIFK gene is not in the TSS data. For now, we simply removed this gene from our analysis, but may explore this more later. \n",
    "\n",
    "We are also checking if any of the entires in the activity_df are of duplicated genes in the TSS annotation (in which case transcript is required to specify which one). For now, we also removed the 75 which were duplicated, and we hope to revisit this. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
